File: src/main/java/de/saschahlusiak/freebloks/client/GameClient.java
package de.saschahlusiak.freebloks.client;

import android.bluetooth.BluetoothSocket;
import android.content.Context;

import androidx.annotation.AnyThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;
import androidx.annotation.WorkerThread;

import com.crashlytics.android.Crashlytics;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import de.saschahlusiak.freebloks.R;
import de.saschahlusiak.freebloks.game.GameConfiguration;
import de.saschahlusiak.freebloks.model.Board;
import de.saschahlusiak.freebloks.model.Game;
import de.saschahlusiak.freebloks.model.GameMode;
import de.saschahlusiak.freebloks.model.Turn;
import de.saschahlusiak.freebloks.network.MessageReadThread;
import de.saschahlusiak.freebloks.network.MessageWriter;
import de.saschahlusiak.freebloks.network.message.MessageChat;
import de.saschahlusiak.freebloks.network.message.MessageRequestGameMode;
import de.saschahlusiak.freebloks.network.message.MessageRequestHint;
import de.saschahlusiak.freebloks.network.message.MessageRequestPlayer;
import de.saschahlusiak.freebloks.network.message.MessageRequestUndo;
import de.saschahlusiak.freebloks.network.message.MessageRevokePlayer;
import de.saschahlusiak.freebloks.network.message.MessageSetStone;
import de.saschahlusiak.freebloks.network.message.MessageStartGame;
import io.fabric.sdk.android.Fabric;

public class GameClient {
	private static final int CONNECT_TIMEOUT = 10000;
	public static final int DEFAULT_PORT = 59995;

	private Closeable clientSocket;
	public final Game game;
	public final Board board;
	private MessageWriter messageWriter;
	private MessageReadThread readThread;
	private ExecutorService sendExecutor;
	private GameConfiguration config;
	private GameClientMessageHandler gameClientMessageHandler;

	@UiThread
	public GameClient(@Nullable Game game, @NonNull GameConfiguration config) {
		this.config = config;
		if (game == null) {
			Board board = new Board(config.getFieldSize());
			board.startNewGame(GameMode.GAMEMODE_4_COLORS_4_PLAYERS);
			board.setAvailableStones(0, 0, 0, 0, 0);

			this.game = new Game(board);
		} else
			this.game = game;

		this.board = this.game.getBoard();

		clientSocket = null;

		gameClientMessageHandler = new GameClientMessageHandler(this.game);
	}

	@Override
	protected void finalize() throws Throwable {
		disconnect();
		super.finalize();
	}

	public GameConfiguration getConfig() {
		return this.config;
	}

	public boolean isConnected() {
		if (clientSocket == null) return false;
		if (clientSocket instanceof Socket && ((Socket) clientSocket).isClosed()) return false;
		if (clientSocket instanceof BluetoothSocket && !((BluetoothSocket) clientSocket).isConnected()) return false;

		return true;
	}

	public void addObserver(GameEventObserver sci) {
		gameClientMessageHandler.addObserver(sci);
	}

	public void removeObserver(GameEventObserver sci) {
		gameClientMessageHandler.removeObserver(sci);
	}

	/**
	 * Try to establish a TCP connection to the given host and port.
	 * On success will call {@link #connected(Closeable, InputStream, OutputStream)}
	 *
	 * @param context for getStrin()
	 * @param host target hostname
	 * @param port port
	 *
	 * @throws IOException on connection refused
	 */
	@WorkerThread
	public void connect(Context context, String host, int port) throws IOException {
		final Socket socket = new Socket();
		try {
			SocketAddress address;
			if (host == null)
				address = new InetSocketAddress((InetAddress)null, port);
			else
				address = new InetSocketAddress(host, port);

			socket.connect(address, CONNECT_TIMEOUT);
		} catch (IOException e) {
			// translate any IOException to "Connection refused"
			e.printStackTrace();

			throw new IOException(context.getString(R.string.connection_refused));
		}

		connected(socket, socket.getInputStream(), socket.getOutputStream());
	}

	/**
	 * Connection is successful, set up message readers and writers.
	 * Make sure you have observers registered before calling this method.
	 *
	 * @param socket a closeable socket, for disconnecting
	 * @param is the InputStream from the socket
	 * @param os the OutputStream to the socket
	 */
	public synchronized void connected(@NonNull Closeable socket, @NonNull InputStream is, @NonNull OutputStream os) {
		this.clientSocket = socket;

		// first we set up writing to the server
		messageWriter = new MessageWriter(os);
		sendExecutor = Executors.newSingleThreadExecutor();

		// we start reading from the server, we will likely begin receiving and processing
		// messages and sending events to the observers.
		// To allow for other observers to be registered before we consume messages, we inform the
		// observers so far about it.
		gameClientMessageHandler.onConnected(this);

		readThread = new MessageReadThread(is, gameClientMessageHandler, () -> {
			disconnect();
			return null;
		});
		readThread.start();
	}

	public synchronized void disconnect() {
		if (clientSocket != null) {
			final Exception lastError = readThread.getError();
			try {
				if (Fabric.isInitialized()) {
					Crashlytics.log("Disconnecting from " + config.getServer());
				}
				readThread.goDown();

				if (clientSocket instanceof Socket) {
					final Socket socket = (Socket) clientSocket;
					if (socket.isConnected())
						socket.shutdownInput();
				}

				if (sendExecutor != null) {
					sendExecutor.shutdown();
					try {
						sendExecutor.awaitTermination(200, TimeUnit.MILLISECONDS);
					} catch (InterruptedException e) {
						e.printStackTrace();
					}
				}
				clientSocket.close();
				readThread = null;
				sendExecutor = null;
			} catch (IOException e) {
				e.printStackTrace();
			}

			gameClientMessageHandler.onDisconnected(this, lastError);
		}
		clientSocket = null;
	}

	/**
	 * Request a new player from the server
	 * @param player player to request
	 * @param name name for the player
	 */
	public void requestPlayer(int player, @Nullable String name) {
		// be aware that the name may be capped at length 16
		send(new MessageRequestPlayer(player, name));
	}

	/**
	 * Request to revoke the given player
	 * @param player the local player to revoke
	 */
	public void revokePlayer(int player) {
		if (!game.isLocalPlayer(player)) return;
		send(new MessageRevokePlayer(player));
	}

	/**
	 * Request a new game mode with new board sizes from the server.
	 *
	 * @param width new width to request
	 * @param height new height to request
	 * @param gameMode new game mode to request
	 * @param stones availability of the 21 stones
	 */
	public void requestGameMode(int width, int height, @NonNull GameMode gameMode, @NonNull int[] stones) {
		send(new MessageRequestGameMode(width, height, gameMode, stones));
	}

	/**
	 * Request a hint for the current local player.
	 */
	public void requestHint() {
		if (game == null)
			return;
		if (!isConnected())
			return;
		if (!game.isLocalPlayer())
			return;
		send(new MessageRequestHint(game.getCurrentPlayer()));
	}

	/**
	 * Send a chat message to the server, which will relay it back
	 * @param message the message
	 */
	public void sendChat(@NonNull String message) {
		// the client does not matter, it will be filled in by the server then broadcasted to all clients
		send(new MessageChat(0, message));
	}

	/**
	 * Called by the UI for the local player to place the stone.
	 * The request is sent to the server, the stone will not be placed locally.
	 */
	public void setStone(@NonNull Turn turn) {
		// locally set no player as the current player
		// on success the server will send us the new current player
		game.setCurrentPlayer(-1);

		send(new MessageSetStone(turn));
	}

	/**
	 * Request server to start the game
	 */
	public void requestGameStart() {
		send(new MessageStartGame());
	}

	/**
	 * Request server to undo the last move
	 */
	public void requestUndo() {
		if (game == null)
			return;
		if (!isConnected())
			return;
		if (!game.isLocalPlayer())
			return;
		send(new MessageRequestUndo());
	}

	/**
	 * Relay the given message to the sendExecutor to be sent to the server asynchronously.
	 * Write errors will be silently ignored.
	 *
	 * @param msg the message to send
	 */
	@AnyThread
	private void send(@NonNull de.saschahlusiak.freebloks.network.Message msg) {
		if (sendExecutor == null)
			return;

		// ignore sending to closed stream
		sendExecutor.submit(() -> messageWriter.write(msg));
	}
}

File: src/main/java/de/saschahlusiak/freebloks/client/GameClient.kt
package de.saschahlusiak.freebloks.client

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.content.Context
import androidx.annotation.AnyThread
import androidx.annotation.UiThread
import de.saschahlusiak.freebloks.R
import de.saschahlusiak.freebloks.crashReporter
import de.saschahlusiak.freebloks.network.bluetooth.BluetoothServerThread
import de.saschahlusiak.freebloks.model.GameConfig
import de.saschahlusiak.freebloks.model.Game
import de.saschahlusiak.freebloks.model.GameMode
import de.saschahlusiak.freebloks.model.Turn
import de.saschahlusiak.freebloks.network.Message
import de.saschahlusiak.freebloks.network.MessageReadThread
import de.saschahlusiak.freebloks.network.MessageWriter
import de.saschahlusiak.freebloks.network.message.*
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.sendBlocking
import java.io.Closeable
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket

// Extend Object so we can override finalize()
@Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
class GameClient @UiThread constructor(@JvmField val game: Game, val config: GameConfig): Object() {

    private var clientSocket: Closeable? = null
    private var readThread: MessageReadThread? = null
    private val gameClientMessageHandler = GameClientMessageHandler(game)

    private val sendQueue = Channel<Message>(Channel.UNLIMITED)
    private var sender: Job? = null

    /**
     * The last status message received by our handler
     */
    val lastStatus get() = gameClientMessageHandler.lastStatus

    @Throws(Throwable::class)
    override fun finalize() {
        disconnect()
        super.finalize()
    }

    fun isConnected(): Boolean {
        return when (val socket = clientSocket) {
            null -> false

            is Socket -> !socket.isClosed
            is BluetoothSocket -> socket.isConnected

            else -> true
        }
    }

    fun addObserver(sci: GameEventObserver) = gameClientMessageHandler.addObserver(sci)
    fun removeObserver(sci: GameEventObserver) = gameClientMessageHandler.removeObserver(sci)

    /**
     * Try to establish a TCP connection to the given host and port.
     * On success will call [.connected]
     *
     * @param context for getString()
     * @param host target hostname or null for localhost
     * @param port port
     *
     * @throws IOException on connection refused
     */
    @UiThread
    suspend fun connect(context: Context, host: String?, port: Int): Boolean {
        val socket = Socket()
        val error = runInterruptible(Dispatchers.IO) {
            val address = host?.let { InetSocketAddress(it, port) }
                ?: InetSocketAddress(null as InetAddress?, port)

            try {
                socket.connect(address, CONNECT_TIMEOUT)
                null
            } catch (e: IOException) {
                e.printStackTrace()

                // translate any IOException to "Connection refused"
                IOException(context.getString(R.string.connection_refused), e)
            }
        }

        if (error != null) {
            gameClientMessageHandler.notifyConnectionFailed(this, error)
            return false
        }

        val (input, output) = runInterruptible(Dispatchers.IO) {
            socket.getInputStream() to socket.getOutputStream()
        }

        connected(socket, input, output)
        return true
    }

    /**
     * Try to connect to the given remote bluetooth device
     *
     * On success will call [.connected]
     */
    suspend fun connect(context: Context, remote: BluetoothDevice): Boolean {
        val (socket, error) = runInterruptible(Dispatchers.IO) {
            try {
                remote.createInsecureRfcommSocketToServiceRecord(BluetoothServerThread.SERVICE_UUID).also {
                    it.connect()
                } to null
            } catch (e: IOException) {
                e.printStackTrace()

                // translate any IOException to "Connection refused"
                null to IOException(context.getString(R.string.connection_refused), e)
            }
        }
        if (error != null) {
            gameClientMessageHandler.notifyConnectionFailed(this, error)
            return false
        }
        socket ?: return false

        connected(socket, socket.inputStream, socket.outputStream)
        return true
    }

    /**
     * Connection is successful, set up message readers and writers.
     * Make sure you have observers registered before calling this method.
     *
     * @param socket a closeable socket, for disconnecting
     * @param input the InputStream from the socket
     * @param output the OutputStream to the socket
     */
    @Synchronized
    @UiThread
    fun connected(socket: Closeable, input: InputStream, output: OutputStream) {
        clientSocket = socket

        // first we set up writing to the server
        sender = GlobalScope.launch(Dispatchers.IO) {
            val writer = MessageWriter(output)
            for (message in sendQueue) writer.write(message)
        }

        // we start reading from the server, we will likely begin receiving and processing
        // messages and sending events to the observers.
        // To allow for other observers to be registered before we consume messages, we inform the
        // observers so far about it.
        gameClientMessageHandler.notifyConnected(this)

        readThread = MessageReadThread(input, gameClientMessageHandler) {
            // this current thread is interrupted, so not a good idea to do much from it
            GlobalScope.launch(Dispatchers.Main) {
                disconnect()
            }
        }.also { it.start() }
    }

    @Synchronized
    @UiThread
    fun disconnect() {
        val socket = clientSocket ?: return

        val lastError = readThread?.error
        try {
            crashReporter.log("Disconnecting from ${config.server}")

            readThread?.goDown()
            sendQueue.close()

            // FIXME: remove runBlocking
            runBlocking {
                withTimeoutOrNull(200) {
                    sender?.join()
                }
            }

            if (socket is Socket && socket.isConnected) {
                socket.shutdownInput()
            }
            socket.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        readThread = null
        sender = null
        clientSocket = null

        gameClientMessageHandler.notifyDisconnected(this, lastError)
    }

    /**
     * Request a new player from the server. This can be called before the client is connected.
     *
     * @param player player to request or -1 for random
     * @param name name for the player
     */
    fun requestPlayer(player: Int, name: String?) {
        // be aware that the name may be capped at length 16
        send(MessageRequestPlayer(player, name))
    }

    /**
     * Request to revoke the given player
     * @param player the local player to revoke
     */
    fun revokePlayer(player: Int) {
        if (!game.isLocalPlayer(player)) return
        send(MessageRevokePlayer(player))
    }

    /**
     * Request a new game mode with new board sizes from the server.
     *
     * @param width new width to request
     * @param height new height to request
     * @param gameMode new game mode to request
     * @param stones availability of the 21 stones
     */
    fun requestGameMode(width: Int, height: Int, gameMode: GameMode, stones: IntArray) {
        send(MessageRequestGameMode(width, height, gameMode, stones))
    }

    /**
     * Request a hint for the current local player.
     */
    fun requestHint() {
        if (!isConnected()) return
        if (!game.isLocalPlayer()) return
        send(MessageRequestHint(game.currentPlayer))
    }

    /**
     * Send a chat message to the server, which will relay it back
     * @param message the message
     */
    fun sendChat(message: String) { // the client does not matter, it will be filled in by the server then broadcasted to all clients
        send(MessageChat(0, message))
    }

    /**
     * Called by the UI for the local player to place the stone.
     * The request is sent to the server, the stone will not be placed locally.
     */
    fun setStone(turn: Turn) {
        // locally set no player as the current player
        // on success the server will send us the new current player
        game.currentPlayer = -1

        send(MessageSetStone(turn))
    }

    /**
     * Request server to start the game. This can be called before the client is connected.
     */
    fun requestGameStart() {
        send(MessageStartGame())
    }

    /**
     * Request server to undo the last move
     */
    fun requestUndo() {
        if (!isConnected()) return
        if (!game.isLocalPlayer()) return
        send(MessageRequestUndo())
    }

    /**
     * Queue the given message to be sent via the [sendQueue]. This can be done even when the [sender] is not running yet.
     *
     * Write errors will be silently ignored.
     *
     * @param msg the message to send
     */
    @AnyThread
    private fun send(msg: Message) {
        sendQueue.sendBlocking(msg)
    }

    companion object {
        private const val CONNECT_TIMEOUT = 5000
        const val DEFAULT_PORT = 59995
    }
}